### PR TITLE
Added more 1.13 data

### PR DESCRIPTION
**Pull request by lluiscab**
Added more 1.13 data. See #200 

- [x] blocks
- [x] biomes
- [x] effects
- [x] enchantments
- [x] entities
- [x] instruments
- [x] items
- [x] language
- [x] materials
- [x] protocol (Was already done before this PR)
- [x] recipes
- [x] windows

### CODE REVIEW

**Comment by lluiscab** [comment]
Tests fail because I've added new values for pretty much everything. Possible block states, minecraft text ids (ex: "acacia_door")

**Comment by lluiscab** [comment]
> However are you sure you have all the information included in 1.12 ?
What do you mean? 1.13 changed how block and items are stored so I'm pretty sure little to no data could be used from the 1.12 spec, could be wrong on that tho.

**Comment by rom1504** [comment]
No I mean the 1.13 files should contain all the fields that were present in
1.12 files. No regression.

**Comment by lluiscab** [comment]
See my PR on burger-extractor, I added some new 1.13 fields but the previous ones are still present

**Comment by rom1504** [comment]
See
https://github.com/PrismarineJS/minecraft-data/blob/master/data/pc/1.12/blocks.json
Vs
https://github.com/lluiscab/minecraft-data/blob/154fbb10fd0722ebb74a5ae736c6f42debffb795/data/pc/1.13/blocks.json

Fields are missing as said by tests.

**Comment by lluiscab** [comment]
Okay I see, I'll look into that tomorrow

**Comment by lluiscab** [comment]
It's finally done, all data has been updated. Waiting for a review now @rom1504

**Comment by rom1504** [comment]
Can you resolve the conflicts ? (git fetch && git rebase origin/master)

File: schemas/blocks_schema.json [hunk 1]
```diff
@@ -76,53 +80,91 @@
...
-        }
+            "required": [
+              "name",
+              "type",
+              "num_values"
+            ],
+            "additionalProperties": false
+          }
+      },
+      "drops": {
```
Comment (rom1504) [inline schemas/blocks_schema.json]: why did you change that ?

File: schemas/items_schema.json [hunk 2]
```diff
@@ -12,6 +12,10 @@
         "type": "integer",
         "minimum": 0
       },
+      "textId": {
```
Comment (rom1504) [inline schemas/items_schema.json]: isn't that the same as name ?

File: schemas/recipes_schema.json [hunk 3]
```diff
@@ -2,11 +2,10 @@
   "$schema": "http://json-schema.org/draft-04/schema#",
   "title": "recipes",
   "description":"A dictionary of quoted numerical item IDs. Each ID maps to a list of recipes. There may be multiple different recipes per item (same ID and metadata). The recipes may not be sorted.",
-
   "definitions": {
     "id": {
-      "type":["integer","null"],
-      "description":"A single numerical ID or null."
+      "type":["string", "integer","null"],
```
Comment (rom1504) [inline schemas/recipes_schema.json]: why not just continue using integer ids ?

File: schemas/recipes_schema.json [hunk 4]
```diff
@@ -31,14 +30,26 @@
       },
       "required": ["id"],
       "additionalProperties":false
-    },
+	},
+	"textId_count":{
```
Comment (rom1504) [inline schemas/recipes_schema.json]: why is that needed ?

File: schemas/blocks_schema.json [hunk 5]
```diff
@@ -12,6 +12,10 @@
         "type": "integer",
         "minimum": 0
       },
+      "textId": {
```
[the remark this answers is not in the document]
Comment (lluiscab) [reply schemas/blocks_schema.json]: It should be and it is for pretty much all the blocks that I've seen, but burger extracts them separately which means they're not the same for the game. So I guess we should keep both values

File: schemas/recipes_schema.json (hunk 3 above)
Comment (lluiscab) [reply schemas/recipes_schema.json]: The game doesn't anymore, why should we?

Comment (rom1504) [reply schemas/recipes_schema.json]: yes it does ? the protocol still has integer ids everywhere.

File: schemas/blocks_schema.json (hunk 5 above)
[the remark this answers is not in the document]
Comment (rom1504) [reply schemas/blocks_schema.json]: "burger extracts them separately which means they're not the same for the game" are you sure ?

File: schemas/blocks_schema.json [hunk 6]
```diff
@@ -27,7 +31,7 @@
           "number",
           "null"
         ],
-        "minimum": 0
+        "minimum": -1
```
[the remark this answers is not in the document]
Comment (lluiscab) [reply schemas/blocks_schema.json]: Unbreakable blocks have a hardness of -1 according to the wiki (See [command block](https://minecraft.gamepedia.com/Command_Block) or the [block breaking trivia](https://minecraft.gamepedia.com/Breaking#Trivia))

File: schemas/blocks_schema.json (hunk 1 above)
Comment (lluiscab) [reply schemas/blocks_schema.json]: Change what? Not really sure what' you're talking about here

File: schemas/blocks_schema.json [hunk 7]
```diff
@@ -146,7 +188,6 @@
       "displayName",
       "name",
       "hardness",
-      "stackSize",
```
[the remark this answers is not in the document]
Comment (lluiscab) [reply schemas/blocks_schema.json]: I'll make a change to the burger-extractor in order to extract this value as well

File: schemas/recipes_schema.json (hunk 3 above)
Comment (lluiscab) [reply schemas/recipes_schema.json]: Okay, I get it. But for recipes, does it really matter?

File: schemas/recipes_schema.json (hunk 4 above)
Comment (lluiscab) [reply schemas/recipes_schema.json]: Metadata is no longer required, so a recipe output that just used a textid and count as params was needed

File: schemas/items_schema.json (hunk 2 above)
Comment (lluiscab) [reply schemas/items_schema.json]: It's probably the same that's happening with the block schema

File: schemas/blocks_schema.json (hunk 1 above)
Comment (rom1504) [reply schemas/blocks_schema.json]: The drop schema

File: schemas/items_schema.json (hunk 2 above)
Comment (rom1504) [reply schemas/items_schema.json]: Yes

File: schemas/blocks_schema.json (hunk 6 above)
[the remark this answers is not in the document]
Comment (lluiscab) [reply schemas/blocks_schema.json]: In 1.12, barrier hardness is set to `null`

File: schemas/recipes_schema.json (hunk 3 above)
Comment (rom1504) [reply schemas/recipes_schema.json]: Not sure what you mean. No it doesn't matter to use one or the other.
However it matters to have consistent files between versions or else it means having separate implementations for each version which is not great.
That doesn't seem too hard to convert recipes to integer ids ?

File: schemas/blocks_schema.json (hunk 1 above)
Comment (lluiscab) [reply schemas/blocks_schema.json]: I just made it possible for the drop array to contain strings for textIds, not really sure why github is saying I changed that much, maybe it's a spacing problem

File: schemas/recipes_schema.json (hunk 4 above)
Comment (rom1504) [reply schemas/recipes_schema.json]: Ok that will solve itself once that is switched back to integer ids.

File: schemas/recipes_schema.json (hunk 3 above)
Comment (lluiscab) [reply schemas/recipes_schema.json]: The game is now using text ids, and so was in 1.12, using text ids makes the files way more readable. Also the recipes don't use numeric ids anywhere in the vanilla game

**Comment by rom1504** [comment]
Ok I will answer more in details to your point but just a general remark :
if we use the textid instead of the integers then we have to release major
versions of every prismarine package and clearly advertise that we are
breaking all the APIs ans everybody need to change their code if they want
to use 1.13

That's what I would like to try to avoid.

File: schemas/blocks_schema.json (hunk 5 above)
[the remark this answers is not in the document]
Comment (rom1504) [reply schemas/blocks_schema.json]: Yes Minecraft data name is what you called text id

File: schemas/blocks_schema.json (hunk 6 above)
[the remark this answers is not in the document]
Comment (rom1504) [reply schemas/blocks_schema.json]: Ok what about we continue using null instead of -1 ?

File: schemas/blocks_schema.json (hunk 1 above)
Comment (rom1504) [reply schemas/blocks_schema.json]: Ok that can be solved by using integer ids

File: schemas/recipes_schema.json (hunk 3 above)
Comment (rom1504) [reply schemas/recipes_schema.json]: Good point but really that file is not highly readable anyway so I'm not sure if we should change this...

File: schemas/blocks_schema.json (hunk 5 above)
[the remark this answers is not in the document]
Comment (lluiscab) [reply schemas/blocks_schema.json]: I have updated the burger extractor so it does not extract the textId anymore. I'll add the changes to this PR soon

File: schemas/blocks_schema.json (hunk 6 above)
[the remark this answers is not in the document]
Comment (lluiscab) [reply schemas/blocks_schema.json]: We could do that but why? Wouldn't it be better to have the value be always an integer?

File: schemas/blocks_schema.json (hunk 1 above)
Comment (lluiscab) [reply schemas/blocks_schema.json]: So block drops should go back to using the block id?

File: schemas/blocks_schema.json (hunk 6 above)
[the remark this answers is not in the document]
Comment (rom1504) [reply schemas/blocks_schema.json]: It's kind of the same thing so no real be benefit, however any change is a drawback because code using that need to be changed.

File: schemas/biomes_schema.json [hunk 8]
```diff
@@ -12,6 +12,10 @@
         "type": "integer",
         "minimum": 0
       },
+      "textId": {
```
Comment (rom1504) [inline schemas/biomes_schema.json]: That's the same as name isn't it ?

Comment (lluiscab) [reply schemas/biomes_schema.json]: It probably is, will remove it if so.

File: schemas/items_schema.json (hunk 2 above)
[the remark this answers is not in the document]
Comment (lluiscab) [reply schemas/items_schema.json]: Yes, this value is still here, just need to get some time to remove it

File: schemas/blocks_schema.json [hunk 9]
```diff
@@ -27,7 +27,7 @@
           "number",
           "null"
         ],
-        "minimum": 0
+        "minimum": -1
```
[the remark this answers is not in the document]
Comment (lluiscab) [reply schemas/blocks_schema.json]: It's still there and I don't see why this shouldn't be changed to always a number, it's easier for type based coding and it's also cleaner

[the remark this answers is not in the document]
Comment (rom1504) [reply schemas/blocks_schema.json]: in that case, it needs to be changed to all other versions and everything using that value (in mineflayer and flying-squid at least)

File: schemas/items_schema.json (hunk 2 above)
[the remark this answers is not in the document]
Comment (rom1504) [reply schemas/items_schema.json]: ok

File: schemas/blocks_schema.json (hunk 9 above)
[the remark this answers is not in the document]
Comment (rom1504) [reply schemas/blocks_schema.json]: it's not a question of whether it's better, but whether it's worth the effort to change it.

[the remark this answers is not in the document]
Comment (rom1504) [reply schemas/blocks_schema.json]: I'd say if you want to change that, it would be simpler to do it in another PR or this is going to get complex

File: schemas/biomes_schema.json (hunk 8 above)
Comment (lluiscab) [reply schemas/biomes_schema.json]: For biomes, the name and textId are not the same, name is actually the display name of the biome (Ex: Frozen Ocean) and the textId is well, the id (ex: frozen_ocean)

File: schemas/items_schema.json (hunk 2 above)
[the remark this answers is not in the document]
Comment (lluiscab) [reply schemas/items_schema.json]: textId for items has been removed

File: schemas/blocks_schema.json (hunk 9 above)
[the remark this answers is not in the document]
Comment (lluiscab) [reply schemas/blocks_schema.json]: Fixed

File: schemas/biomes_schema.json (hunk 8 above)
Comment (rom1504) [reply schemas/biomes_schema.json]: Ok. Could you call textid name and what is currently name would become "displayName" ?
That won't break anything since we don't use biomes that much.

Comment (lluiscab) [reply schemas/biomes_schema.json]: Yeah expect that changing the schema to add displayName will basically make all the biomes.json prior to 1.13 fail

Comment (rom1504) [reply schemas/biomes_schema.json]: Just make it optional

Comment (lluiscab) [reply schemas/biomes_schema.json]: Fixed!

File: schemas/recipes_schema.json (hunk 3 above)
Comment (lluiscab) [reply schemas/recipes_schema.json]: Fixed

File: schemas/recipes_schema.json [hunk 10]
```diff
@@ -2,11 +2,10 @@
   "$schema": "http://json-schema.org/draft-04/schema#",
   "title": "recipes",
   "description":"A dictionary of quoted numerical item IDs. Each ID maps to a list of recipes. There may be multiple different recipes per item (same ID and metadata). The recipes may not be sorted.",
-
   "definitions": {
     "id": {
       "type":["integer","null"],
-      "description":"A single numerical ID or null."
+      "description":"A single numerical ID, text id or null."
```
Comment (rom1504) [inline schemas/recipes_schema.json]: Remove that from the schema now that IDs are all int

File: schemas/recipes_schema.json [hunk 11]
```diff
@@ -89,11 +88,10 @@
       ]
     }
   },
-
   "type": "object",
   "uniqueItems": true,
   "patternProperties" : {
-    "^[0-9]+$": {
+    "^[0-9 A-z]+$": {
```
Comment (rom1504) [inline schemas/recipes_schema.json]: Remove that now that IDs are all int

File: schemas/recipes_schema.json (hunk 10 above)
Comment (lluiscab) [reply schemas/recipes_schema.json]: Done

File: schemas/recipes_schema.json (hunk 11 above)
Comment (lluiscab) [reply schemas/recipes_schema.json]: Done

**Comment by lluiscab** [comment]
Everything should be ready

File: data/dataPaths.json [hunk 12]
```diff
@@ -339,12 +339,12 @@
       "language":"pc/1.12"
     },
     "1.13": {
```
[the remark this answers is not in the document]
Comment (lluiscab) [reply data/dataPaths.json]: Will fix tomorrow, just a typo when I rebased

**Comment by rom1504** [comment]
Things still missing : 
* in biomes : nothing
* in blocks : 
comparing https://github.com/PrismarineJS/minecraft-data/blob/master/data/pc/1.12/blocks.json vs 
https://github.com/PrismarineJS/minecraft-data/blob/bdd4ff1c83680c8bf00aabce16657644a702ba6d/data/pc/1.13/blocks.json
  * [ ] harvestTools : sure it's an optional property, but it's very useful for bots to know what tool to use (+ it's available in the wiki)
  * [ ] variations : is that completely related by new blocks now ?
  * [ ] material : I believe that's only available in the wiki but this is required for the digging time computation algo which is required to dig a block in mineflayer
* in effects : nothing
* in enchantments : nothing
* in entities : nothing
* in instruments : nothing
* in items : nothing
* in language : nothing
* [ ] material.json is missing
* in protocol : nothing
* in recipes : looks ok but honestly it's hard to be sure
* in version : nothing
* in windows : nothing

Things that really should be fixed : 
* [ ] point all files to 1.13 in datapath
* [ ] copy material.json
* [ ] material property in blocks.json

Optional but would be nice
* [ ] harvestTools : not used in mineflayer, but very useful for bots

**Comment by lluiscab** [comment]
Things that should be fixed:

- [x] point all files to 1.13 in datapath
- [x] copy material.json
- [ ] material property in blocks.json
- [ ] harvestTools property in blocks.json

**Comment by rom1504** [comment]
the harvestTools value and the content of materials.json are wrong.
Reading the burger extractor code, I'm under the impression it is just copying the old (1.12) data if it cannot find it.
I can't see how that can produce up to date data.

**Comment by lluiscab** [comment]
Yikes for the id change, will see if I can update that tomorrow

**Comment by lluiscab** [comment]
Should be all ready now (Hope this time is true)

**Comment by rom1504** [comment]
looks good, merged, thanks for all the work on this !
